dask/layers.py

```diff
@@ -100,6 +102,235 @@ def __init__(
         )
 
 
+class ArrayOverlapLayer(Layer):
+    """Simple HighLevelGraph array overlap layer.
+
+    Lazily computed High-level graph layer for a array overlap operations.
+
+    Parameters
+    ----------
+    name : str
+        Name of new output overlap array.
+    array : Dask array
+    axes: Mapping
+        Axes dictionary indicating overlap in each dimension,
+        e.g. ``{'0': 1, '1': 1}``
+    """
+
+    def __init__(
+        self,
+        name,
+        axes,
+        chunks,
+        numblocks,
+        token,
+    ):
+        super().__init__()
+        self.name = name
+        self.axes = axes
+        self.chunks = chunks
+        self.numblocks = numblocks
+        self.token = token
+        self._cached_keys = None
+
+    def __repr__(self):
+        return "ArrayOverlapLayer<name='{}'".format(self.name)
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def is_materialized(self):
+        return hasattr(self, "_cached_dict")
+
+    def _dask_keys(self):
+        if self._cached_keys is not None:
+            return self._cached_keys
+
+        name, chunks, numblocks = self.name, self.chunks, self.numblocks
+
+        def keys(*args):
+            if not chunks:
+                return [(name,)]
+            ind = len(args)
+            if ind + 1 == len(numblocks):
+                result = [(name,) + args + (i,) for i in range(numblocks[ind])]
+            else:
+                result = [keys(*(args + (i,))) for i in range(numblocks[ind])]
+            return result
+
+        self._cached_keys = result = keys()
+        return result
+
+    def _construct_graph(self, deserializing=False):
+        """Construct graph for a simple overlap operation."""
+        axes = self.axes
+        chunks = self.chunks
+        name = self.name
+        dask_keys = self._dask_keys()
+
+        getitem_name = "getitem-" + self.token
+        overlap_name = "overlap-" + self.token
+
+        if deserializing:
+            # Use CallableLazyImport objects to avoid importing dataframe
+            # module on the scheduler
+            concatenate3 = CallableLazyImport("dask.array.core.concatenate3")
+        else:
+            # Not running on distributed scheduler - Use explicit functions
+            from dask.array.core import concatenate3
+
+        dims = list(map(len, chunks))
+        expand_key2 = toolz.partial(expand_key, dims=dims, axes=axes)
```

Comment:
Nitpick: let's use Python's built-in `partial` function here (`toolz.partial` is just an alias for the built-in `partial` anyways)

```python
In [1]: from functools import partial

In [2]: from functools import partial as tpartial

In [3]: tpartial is partial
Out[3]: True
```